File: src-tauri/src/commands/adjuntos.rs
//! Commands of `adjuntos`. See `docs/11-contratos-tauri.md` §5.12.

use certaro_application::dtos::adjuntos::{AdjuntoInput, AdjuntoItem};
use certaro_domain::entities::EntidadAdjunto;
use tauri::State;
use uuid::Uuid;

use crate::error::{handle, ApiResult};
use crate::state::AppState;

#[tauri::command]
pub async fn adjuntos_list(
    state: State<'_, AppState>,
    entidad_tipo: EntidadAdjunto,
    entidad_id: Uuid,
) -> ApiResult<Vec<AdjuntoItem>> {
    let outcome = match state.services() {
        Ok(services) => services.adjuntos.list(entidad_tipo, entidad_id).await,
        Err(e) => Err(e),
    };
    handle("adjuntos_list", outcome)
}

#[tauri::command]
pub async fn adjuntos_add(
    state: State<'_, AppState>,
    input: AdjuntoInput,
) -> ApiResult<AdjuntoItem> {
    let outcome = match state.services() {
        Ok(services) => services.adjuntos.add(input).await,
        Err(e) => Err(e),
    };
    handle("adjuntos_add", outcome)
}

#[tauri::command]
pub async fn adjuntos_delete(state: State<'_, AppState>, id: Uuid) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.adjuntos.delete(id).await,
        Err(e) => Err(e),
    };
    handle("adjuntos_delete", outcome)
}

#[tauri::command]
pub async fn adjuntos_open(state: State<'_, AppState>, id: Uuid) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.adjuntos.open(id).await,
        Err(e) => Err(e),
    };
    handle("adjuntos_open", outcome)
}

#[tauri::command]
pub async fn adjuntos_reveal(state: State<'_, AppState>, id: Uuid) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.adjuntos.reveal(id).await,
        Err(e) => Err(e),
    };
    handle("adjuntos_reveal", outcome)
}

File: src/api/adjuntos.ts
import { callCommand } from './client'
import type { Uuid } from './types'

/** See `docs/11-contratos-tauri.md` §5.12 and `docs/13-servicios-externos-y-archivos.md` §1. */

export type EntidadAdjunto =
  | 'Movimiento'
  | 'Factura'
  | 'Certificado'
  | 'Liquidacion'
  | 'Cliente'
  | 'Proyecto'
  | 'Trabajo'

export interface AdjuntoItem {
  id: Uuid
  entidadTipo: EntidadAdjunto
  entidadId: Uuid
  nombreArchivo: string
  mime: string
  /** Bytes. The component turns this into KB or MB. */
  tamano: number
  adjuntadoEn: string
}

export interface AdjuntoInput {
  entidadTipo: EntidadAdjunto
  entidadId: Uuid
  rutaOrigen: string
}

export function listAdjuntos(
  entidadTipo: EntidadAdjunto,
  entidadId: Uuid,
): Promise<AdjuntoItem[]> {
  return callCommand('adjuntos_list', { entidadTipo, entidadId })
}

export function addAdjunto(input: AdjuntoInput): Promise<AdjuntoItem> {
  return callCommand('adjuntos_add', { input })
}

export function deleteAdjunto(id: Uuid): Promise<void> {
  return callCommand('adjuntos_delete', { id })
}

export function openAdjunto(id: Uuid): Promise<void> {
  return callCommand('adjuntos_open', { id })
}

export function revealAdjunto(id: Uuid): Promise<void> {
  return callCommand('adjuntos_reveal', { id })
}
